op-node/rollup/derive/frame_queue_test.go

```diff
@@ -0,0 +1,154 @@
+package derive
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"log/slog"
+	"testing"
+
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum-optimism/optimism/op-node/rollup/derive/mocks"
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum-optimism/optimism/op-service/testlog"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+)
+
+func TestPruneFrameQueue(t *testing.T) {
+	for _, tt := range []struct {
+		desc string
+		fs   []testFrame
+		exp  []testFrame
+	}{
+		{
+			desc: "empty",
+			fs:   []testFrame{},
+			exp:  []testFrame{},
+		},
+		{
+			desc: "one",
+			fs:   []testFrame{"a:2:"},
+			exp:  []testFrame{"a:2:"},
+		},
+		{
+			desc: "one-last",
+			fs:   []testFrame{"a:2:!"},
+			exp:  []testFrame{"a:2:!"},
+		},
+		{
+			desc: "last-new",
+			fs:   []testFrame{"a:2:!", "b:0:"},
+			exp:  []testFrame{"a:2:!", "b:0:"},
+		},
+		{
+			desc: "last-ooo",
+			fs:   []testFrame{"a:2:!", "b:1:"},
+			exp:  []testFrame{"a:2:!"},
+		},
+		{
+			desc: "middle-lastooo",
+			fs:   []testFrame{"b:1:", "a:2:!"},
+			exp:  []testFrame{"b:1:"},
+		},
+		{
+			desc: "middle-first",
+			fs:   []testFrame{"b:1:", "a:0:"},
+			exp:  []testFrame{"a:0:"},
+		},
+		{
+			desc: "last-first",
+			fs:   []testFrame{"b:1:!", "a:0:"},
+			exp:  []testFrame{"b:1:!", "a:0:"},
+		},
+		{
+			desc: "last-ooo",
+			fs:   []testFrame{"b:1:!", "b:2:"},
+			exp:  []testFrame{"b:1:!"},
+		},
+		{
+			desc: "ooo",
+			fs:   []testFrame{"b:1:", "b:3:"},
+			exp:  []testFrame{"b:1:"},
+		},
+		{
+			desc: "other-ooo",
+			fs:   []testFrame{"b:1:", "c:3:"},
+			exp:  []testFrame{"b:1:"},
+		},
+		{
+			desc: "other-ooo-last",
+			fs:   []testFrame{"b:1:", "c:3:", "b:2:!"},
+			exp:  []testFrame{"b:1:", "b:2:!"},
+		},
+		{
+			desc: "first-discards-multiple",
+			fs:   []testFrame{"c:0:", "c:1:", "c:2:", "d:0:", "c:3:!"},
+			exp:  []testFrame{"d:0:"},
+		},
+		{
+			desc: "complex",
+			fs:   []testFrame{"b:1:", "b:2:!", "a:0:", "c:1:!", "a:1:", "a:2:!", "c:0:", "c:1:", "d:0:", "c:2:!", "e:0:"},
```

Comment:
What picks up the case where the first frame for the first channel we read isn't submitted? ie in this case `b:0:` was never submitted. I think it's fine because the channel won't ever be considered complete and ready for processing and eventually a first frame from a channel will be received which invalidates it.

Reply:
This would be handled 
1. in the next stage (channel bank), where if it pulls an out of order frame, it is also immediately dropped. I don't want to remember the last-pulled frame in the frame queue to already prune it here instead.
2. as you say, if a first frame arrives in the frame queue, it would also be pruned already in the frame queue.